File: sdks/node/src/runtime/artifacts.ts
import type PubNub from 'pubnub';
import { DEFAULTS } from '../defaults.js';
import { bytesToBase64 } from '../stream/bytes.js';

export interface DownloadedArtifact {
  data: Uint8Array;
  mimeType: string;
  fileName?: string;
}

export interface ArtifactRef {
  kind: 'inline' | 'file';
  mimeType: string;
  size: number;
  hash?: string;
  /** Base64-encoded data for inline artifacts (kind: 'inline' only) */
  data?: string;
  /** PubNub Files identifier (kind: 'file' only) */
  fileId?: string;
  /** File name (both variants; required for kind: 'file') */
  fileName?: string;
  /** PubNub channel where the file is stored (kind: 'file' only, required).
   *  Used with fileId and fileName for pubnub.downloadFile() under PAM. */
  channel?: string;
  expiresAt?: string;
}

export const shouldInlineArtifact = (
  sizeBytes: number,
  inlineLimit: number = DEFAULTS.inlineLimitBytes,
): boolean => sizeBytes <= inlineLimit;

export interface BuildArtifactInput {
  mimeType: string;
  /** Size in bytes - required if data is not provided */
  size?: number;
  hash?: string;
  /** Raw data for inline artifacts - will be base64 encoded. Accepts
   *  `Uint8Array` (includes Node `Buffer` at runtime). */
  data?: Uint8Array;
  /** File reference for uploaded artifacts. Channel is the PubNub channel
   *  where the file is stored (required for downloadFile under PAM). */
  file?: { id: string; name: string; channel: string; expiresAt?: string };
  /** Original file name (preserved on inline artifacts too). */
  fileName?: string;
  inlineLimit?: number;
}

export const buildArtifactRef = (input: BuildArtifactInput): ArtifactRef => {
  const inlineLimit = input.inlineLimit ?? DEFAULTS.inlineLimitBytes;
  const size = input.size ?? input.data?.length ?? 0;
  const inline = shouldInlineArtifact(size, inlineLimit) || !input.file;

  if (inline) {
    return {
      kind: 'inline',
      mimeType: input.mimeType,
      size,
      hash: input.hash,
      // bytesToBase64 is Buffer-independent and works in both Node and
      // browsers. Uint8Array.toString('base64') returns a comma-joined
      // decimal string on runtimes where Buffer is not defined --
      // silently corrupting artifact data.
      data: input.data !== undefined ? bytesToBase64(input.data) : undefined,
      fileName: input.fileName,
    };
  }

  // file is guaranteed to be defined here since inline is false only when file exists
  const file = input.file!;
  return {
    kind: 'file',
    mimeType: input.mimeType,
    size,
    hash: input.hash,
    fileId: file.id,
    fileName: file.name,
    channel: file.channel,
    expiresAt: file.expiresAt,
  };
};

/**
 * Decode an inline base64 artifact ref into raw bytes.
 * Browser-safe: uses atob() + Uint8Array (no Buffer dependency).
 *
 * The ArtifactRef shape uses top-level `kind: 'inline'` discriminator
 * with `data` as a top-level base64 string field (see ArtifactRef
 * interface and artifact-ref.schema.json).
 */
export function decodeInlineArtifact(ref: ArtifactRef): Uint8Array {
  if (ref.kind !== 'inline' || !ref.data) {
    throw new Error('ArtifactRef is not an inline artifact or has no data');
  }
  const binaryString = atob(ref.data);
  const bytes = new Uint8Array(binaryString.length);
  for (let i = 0; i < binaryString.length; i++) {
    bytes[i] = binaryString.charCodeAt(i);
  }
  return bytes;
}

/**
 * Download an artifact from an ArtifactRef.
 *
 * - Inline artifacts: decode base64 data directly (no PubNub call).
 * - File artifacts: call pubnub.downloadFile() using the ref's
 *   channel, fileId, and fileName.
 *
 * Requires a PubNub instance with a valid token granting READ on the
 * artifact's channel. For TaskSession consumers, the T4 read token
 * already grants this.
 */
export async function downloadArtifact(
  ref: ArtifactRef,
  pubnub: PubNub,
): Promise<DownloadedArtifact> {
  if (ref.kind === 'inline') {
    if (!ref.data) {
      throw new Error('Inline ArtifactRef is missing data');
    }
    const data = decodeInlineArtifact(ref);
    return { data, mimeType: ref.mimeType, fileName: ref.fileName };
  }

  if (ref.kind === 'file') {
    if (!ref.channel || !ref.fileId || !ref.fileName) {
      throw new Error(
        'File ArtifactRef is missing required fields: channel, fileId, and fileName are all required',
      );
    }

    const result = await pubnub.downloadFile({
      channel: ref.channel,
      id: ref.fileId,
      name: ref.fileName,
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const file = (result as any).data ?? result;
    let data: Uint8Array;

    // PubNub SDK v10 returns result.data as a Uint8Array (Node Buffer
    // extends Uint8Array, so the instanceof check catches both Node
    // and browser Uint8Array shapes) or as a Blob in older browser
    // builds. Even older @pubnub/file objects exposed toBuffer() /
    // toArrayBuffer(); we keep duck-typed fallbacks for those.
    if (file instanceof Uint8Array) {
      data = new Uint8Array(file);
    } else if (typeof Blob !== 'undefined' && file instanceof Blob) {
      const ab = await file.arrayBuffer();
      data = new Uint8Array(ab);
    } else if (typeof (file as { toArrayBuffer?: unknown })?.toArrayBuffer === 'function') {
      const ab = await (file as { toArrayBuffer: () => Promise<ArrayBuffer> }).toArrayBuffer();
      data = new Uint8Array(ab);
    } else if (typeof (file as { toBuffer?: unknown })?.toBuffer === 'function') {
      const buf = await (file as { toBuffer: () => Promise<Uint8Array | ArrayBuffer> }).toBuffer();
      data = new Uint8Array(buf);
    } else {
      throw new Error('PubNub downloadFile result has no usable data format');
    }

    return { data, mimeType: ref.mimeType, fileName: ref.fileName };
  }

  throw new Error(`Unknown ArtifactRef kind: ${ref.kind}`);
}

File: sdks/python/blocks_network/artifacts.py
"""
Artifact reference builder and downloader.

Port of ``src/shared/artifacts.ts``.
"""

from __future__ import annotations

import base64
from dataclasses import dataclass
from typing import Any, Dict, Optional, Union

from . import config as _cfg
from .types import ArtifactRef


# ============================================================================
# Downloaded artifact
# ============================================================================


@dataclass
class DownloadedArtifact:
    """Result of downloading an artifact from an ArtifactRef."""

    data: bytes
    mime_type: str
    file_name: Optional[str] = None


# ============================================================================
# Decode / Download helpers
# ============================================================================


def decode_inline_artifact(ref: ArtifactRef) -> bytes:
    """Decode an inline artifact's base64 data to raw bytes.

    Raises ValueError if the ref has no ``data`` field.
    """
    if not ref.data:
        raise ValueError("Inline artifact ref is missing 'data' field")
    return base64.b64decode(ref.data)


def download_artifact(ref: ArtifactRef, pubnub: Any) -> DownloadedArtifact:
    """Download an artifact from an ArtifactRef.

    Inline artifacts are decoded from base64 (no PubNub call).
    File artifacts are downloaded via ``pubnub.download_file()``.

    Requires a PubNub instance with a valid token granting READ on
    the artifact's channel. For TaskSession consumers, the T4 read
    token already grants this.
    """
    if ref.kind == "inline":
        data = decode_inline_artifact(ref)
        return DownloadedArtifact(
            data=data,
            mime_type=ref.mime_type,
            file_name=ref.file_name,
        )

    if ref.kind == "file":
        if not ref.channel or not ref.file_id or not ref.file_name:
            raise ValueError(
                "File artifact ref is missing required fields "
                "(channel, file_id, file_name)"
            )
        result = (
            pubnub.download_file()
            .channel(ref.channel)
            .file_id(ref.file_id)
            .file_name(ref.file_name)
            .sync()
        )
        data = result.result.data
        if not isinstance(data, bytes):
            data = bytes(data)
        return DownloadedArtifact(
            data=data,
            mime_type=ref.mime_type,
            file_name=ref.file_name,
        )

    raise ValueError(f"Unknown artifact kind: {ref.kind!r}")


def should_inline_artifact(
    size_bytes: int,
    inline_limit: Optional[int] = None,
) -> bool:
    """Return ``True`` if an artifact of *size_bytes* should be inlined.

    The default limit comes from
    ``ARTIFACT_INLINE_LIMIT_BYTES`` in :mod:`blocks_network.config`.
    """
    if inline_limit is None:
        inline_limit = _cfg.ARTIFACT_INLINE_LIMIT_BYTES
    return size_bytes <= inline_limit


def build_artifact_ref(
    *,
    data: Optional[Union[bytes, str]] = None,
    mime_type: str = "application/octet-stream",
    size: Optional[int] = None,
    hash: Optional[str] = None,
    file: Optional[Dict[str, str]] = None,
    file_name: Optional[str] = None,
    inline_limit: Optional[int] = None,
) -> Dict[str, Any]:
    """Build an ``ArtifactRef`` dict (camelCase, ready for PubNub publish).

    Parameters
    ----------
    data:
        Raw artifact bytes (or UTF-8 string). Will be base64-encoded for
        inline artifacts.
    mime_type:
        MIME type of the artifact.
    size:
        Explicit size in bytes. Computed from *data* if omitted.
    hash:
        Integrity hash string (e.g. ``"sha256:abc..."``).
    file:
        Dict with keys ``id``, ``name``, ``channel``, and optionally
        ``expires_at`` for file-based artifacts.
    file_name:
        Original filename. Applies to both inline and file variants.
        For file variants, ``file["name"]`` takes precedence if present.
    inline_limit:
        Override for the inline size threshold.

    Returns
    -------
    dict
        A camelCase dict matching the ``ArtifactRef`` JSON schema.
    """
    # Normalize data to bytes
    raw: Optional[bytes] = None
    if isinstance(data, str):
        raw = data.encode("utf-8")
    elif isinstance(data, (bytes, bytearray)):
        raw = bytes(data)

    effective_size = size if size is not None else (len(raw) if raw is not None else 0)
    limit = inline_limit if inline_limit is not None else _cfg.ARTIFACT_INLINE_LIMIT_BYTES

    inline = should_inline_artifact(effective_size, limit) or file is None

    if inline:
        ref = ArtifactRef(
            kind="inline",
            mime_type=mime_type,
            size=effective_size,
            hash=hash,
            data=base64.b64encode(raw).decode("ascii") if raw is not None else None,
            file_name=file_name,
        )
    else:
        ref = ArtifactRef(
            kind="file",
            mime_type=mime_type,
            size=effective_size,
            hash=hash,
            channel=file.get("channel"),
            file_id=file.get("id"),
            file_name=file.get("name") or file_name,
            expires_at=file.get("expires_at") or file.get("expiresAt"),
        )

    return ref.to_dict()
